packages/appium/docs/en/ecosystem/build-plugins.md

```diff
@@ -2,4 +2,436 @@
+    }
+}
+```
+
+Now, assuming your plugin is named `myplugin`, users of your plugin can run `appium plugin run
+myplugin prebuild`, and your script will execute.
+
+### Add new Appium commands
+
+If you want to offer functionality that doesn't map to any of the existing commands supported by
+drivers, you can create new commands in one of two ways, just as is possible for drivers:
+
+1. Extending the WebDriver protocol and creating client-side plugins to access the extensions
+1. Overloading the Execute Script command by defining [Execute
+   Methods](../guides/execute-methods.md)
+
+If you want to follow the first path, you can direct Appium to recognize new methods and add them
+to its set of allowed HTTP routes and command names. You do this by assigning the `newMethodMap`
+static variable in your driver class to an object of the same form as Appium's `routes.js` object.
+For example, here is part of the `newMethodMap` for the `FakePlugin` example driver:
+
+```js
+static newMethodMap = {
+  '/session/:sessionId/fake_data': {
+    GET: {command: 'getFakeSessionData', neverProxy: true},
+    POST: {
+      command: 'setFakeSessionData',
+      payloadParams: {required: ['data']},
+      neverProxy: true,
+    },
+  },
+  '/session/:sessionId/fakepluginargs': {
+    GET: {command: 'getFakePluginArgs', neverProxy: true},
+  },
+};
+```
+
+In this example we're adding a few new routes and a total of 3 new commands. For more examples of
+how to define commands in this way, it's best to have a look through `routes.js`. Now all you need
+to do is implement the command handlers in the same way you'd implement any other Appium command.
+
+Note also the special `neverProxy` key for commands; this is generally a good idea to set to `true`
+for plugins, since your plugin might be active for a driver that is put into proxy mode but hasn't
+bothered to decline proxying for these (new and therefore unknown) commands. Setting `neverProxy`
+to `true` here will cause Appium to never proxy these routes and therefore ensure your plugin
+handles them, even if a driver is in proxy mode.
+
+The downside of adding new commands via `newMethodMap` is that people using the standard Appium
+clients won't have nice client-side functions designed to target these endpoints. So you would need
+to create and release client-side plugins for each language you want to support (directions or
+examples can be found at the relevant client docs).
+
+An alternative to this way of doing things is to overload a command which all WebDriver clients
+have access to already: Execute Script. Make sure to read the section on [adding new
+commands](./build-drivers.md#extend-the-existing-protocol-with-new-commands) in
+the Building Drivers guide to understand the way this works in general. The way it works with
+plugins is only slightly different. Let's look at an example taken from Appium's `fake-plugin`:
+
+```js
+static executeMethodMap = {
```

Comment:
from this example it is unclear the definition should be located inside of the plugin class

Reply:
ok made that more clear.